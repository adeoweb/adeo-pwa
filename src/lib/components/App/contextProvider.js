import React from 'react';
import {
    ToastContextProvider,
    WindowSizeContextProvider
} from '@magento/peregrine';

import { PeregrineContextProvider } from 'src/peregrine/lib';
/**
 * List of context providers that are required to run Venia
 *
 * @property {React.Component[]} contextProviders
 */
const contextProviders = [
    PeregrineContextProvider,
    WindowSizeContextProvider,
    ToastContextProvider
];

const ContextProvider = ({ children }) => {
    return contextProviders.reduceRight((memo, ContextProvider) => {
        return <ContextProvider>{memo}</ContextProvider>;
    }, children);
};

export default ContextProvider;
