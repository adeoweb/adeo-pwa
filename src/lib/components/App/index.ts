export { default } from './Container';
export { default as AppContextProvider } from './ContextProvider';
