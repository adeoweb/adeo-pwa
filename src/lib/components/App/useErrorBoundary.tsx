import React, { ReactNode, useMemo } from 'react';

interface State {
    renderError?: Error;
}

export const useErrorBoundary = (WrappedComponent: ReactNode) =>
    useMemo(
        () =>
            class ErrorBoundary extends React.Component<
                Record<string, never>,
                State
            > {
                constructor(props) {
                    super(props);

                    this.state = { renderError: undefined };
                }

                static getDerivedStateFromError(renderError: Error) {
                    return { renderError };
                }

                render(): ReactNode {
                    const { props, state } = this;
                    const { renderError } = state;

                    return (
                        // @ts-expect-error
                        <WrappedComponent
                            {...props}
                            renderError={renderError}
                        />
                    );
                }
            },
        []
    );
