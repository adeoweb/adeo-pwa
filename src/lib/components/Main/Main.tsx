import React, { Fragment, FunctionComponent } from 'react';
import { useScrollLock } from '@magento/peregrine';

import Footer from '../Footer';
import Header from '../Header';
import MessageCard from 'src/lib/components/MessageCard';

interface IMain {
    isMasked: boolean;
}

const Main: FunctionComponent<IMain> = ({ children, isMasked }) => {
    useScrollLock(isMasked);

    return (
        <Fragment>
            <Header />
            <MessageCard />
            <main className={'main'}>{children}</main>
            <Footer />
        </Fragment>
    );
};

export default Main;
