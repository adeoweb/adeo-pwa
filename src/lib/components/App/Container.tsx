import {
    ApolloProvider as ApolloHooksProvider,
    ApolloClient
} from '@apollo/react-hooks';
import React from 'react';
import { IntlProvider } from 'react-intl';

import { useErrorContext } from 'src/peregrine/lib/context/adeoweb/unhandledErrors';

import App from './app';
import { useErrorBoundary } from './useErrorBoundary';

interface IAppContainerProps {
    apolloClient: ApolloClient<unknown>;
}

const AppContainer = ({ apolloClient }: IAppContainerProps): JSX.Element => {
    const ErrorBoundary = useErrorBoundary(App);
    const [unhandledErrors, errorApi] = useErrorContext();

    return (
        <ApolloHooksProvider client={apolloClient}>
            <IntlProvider locale="en">
                <ErrorBoundary
                    unhandledErrors={unhandledErrors}
                    {...errorApi}
                />
            </IntlProvider>
        </ApolloHooksProvider>
    );
};

export default AppContainer;
