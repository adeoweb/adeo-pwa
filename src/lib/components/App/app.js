require('../../styles/styles.scss');
import React, { useCallback, useEffect } from 'react';
import { array, func, shape, string } from 'prop-types';
import { useToasts } from '@magento/peregrine';
import AlertCircleIcon from 'react-feather/alert-circle';
import CloudOffIcon from 'react-feather/cloud-off';
import WifiIcon from 'react-feather/wifi';
import RefreshIcon from 'react-feather/refresh-ccw';

import { useApp } from 'src/peregrine/lib/talons/adeoweb/App/useApp';
import { useLoadConfig } from 'src/peregrine/lib/talons/adeoweb/App/useLoadConfig';
import { useLoadUser } from 'src/peregrine/lib/talons/adeoweb/User/useLoadUser';
import { useUserSession } from 'src/peregrine/lib/talons/adeoweb/User/useUserSession';
import LoadingOverlay from 'src/lib/components/LoadingOverlay';
import SIGN_OUT_MUTATION from 'src/lib/queries/signOut.graphql';
import { Route, Switch, useHistory } from 'src/lib/drivers';
import CustomerModal from 'src/lib/components/CustomerModal';
import scrollTo from 'src/lib/util/scrollTo';

import GET_STORE_CONFIG_DATA from '../../queries/getStoreConfigData.graphql';
import GET_CUSTOMER_QUERY from '../../queries/getCustomer.graphql';
import { HTML_UPDATE_AVAILABLE } from '../../constants/swMessageTypes';
import { registerMessageHandler } from '../../util/swUtils';
import Checkout, { rootCheckoutRoute } from '../Checkout';
import ToastContainer from '../ToastContainer';
import { HeadProvider, Title } from '../Head';
import CookieWarning from '../CookieWarning';
import MobileMenu from '../MobileMenu';
import Routes from '../Routes';
import Icon from '../Icon';
import Main from '../Main';
import Mask from '../Mask';

const OnlineIcon = <Icon src={WifiIcon} attrs={{ width: 18 }} />;
const OfflineIcon = <Icon src={CloudOffIcon} attrs={{ width: 18 }} />;
const ErrorIcon = <Icon src={AlertCircleIcon} attrs={{ width: 18 }} />;
const UpdateIcon = <Icon src={RefreshIcon} attrs={{ width: 18 }} />;

const ERROR_MESSAGE = 'Sorry! An unexpected error occurred.';

const App = props => {
    const { markErrorHandled, renderError, unhandledErrors } = props;
    const [, { addToast }] = useToasts();

    const handleIsOffline = useCallback(() => {
        addToast({
            type: 'error',
            icon: OfflineIcon,
            message: 'You are offline. Some features may be unavailable.',
            timeout: 3000
        });
    }, [addToast]);

    const handleIsOnline = useCallback(() => {
        addToast({
            type: 'info',
            icon: OnlineIcon,
            message: 'You are online.',
            timeout: 3000
        });
    }, [addToast]);

    const handleHTMLUpdate = useCallback(
        resetHTMLUpdateAvaiableFlag => {
            addToast({
                type: 'warning',
                icon: UpdateIcon,
                message: 'Update available. Please refresh.',
                actionText: 'Refresh',
                timeout: 0,
                onAction: () => {
                    location.reload();
                },
                onDismiss: removeToast => {
                    resetHTMLUpdateAvaiableFlag();
                    removeToast();
                }
            });
        },
        [addToast]
    );

    const handleError = useCallback(
        (error, id, loc, handleDismissError) => {
            const errorToastProps = {
                icon: ErrorIcon,
                message: `${ERROR_MESSAGE}\nDebug: ${id} ${loc}`,
                onDismiss: remove => {
                    handleDismissError();
                    remove();
                },
                timeout: 15000,
                type: 'error'
            };

            addToast(errorToastProps);
        },
        [addToast]
    );

    const talonProps = useApp({
        handleError,
        handleIsOffline,
        handleIsOnline,
        handleHTMLUpdate,
        markErrorHandled,
        renderError,
        unhandledErrors
    });

    const {
        hasOverlay,
        handleCloseDrawer,
        setHTMLUpdateAvailable,
        hideCustomerModal,
        customerModalType
    } = talonProps;

    const { error: configError, loaded } = useLoadConfig({
        getConfigQuery: GET_STORE_CONFIG_DATA
    });

    const { error: userDetailsError } = useLoadUser({
        getUserDetailsQuery: GET_CUSTOMER_QUERY
    });

    useUserSession({ signOutMutation: SIGN_OUT_MUTATION });

    useEffect(() => {
        const unregisterHandler = registerMessageHandler(
            HTML_UPDATE_AVAILABLE,
            () => {
                setHTMLUpdateAvailable(true);
            }
        );
        return unregisterHandler;
    }, [setHTMLUpdateAvailable]);

    useEffect(() => {
        document.body.classList.add('loaded');
    }, [loaded]);

    const { listen: addHistoryListener } = useHistory();

    useEffect(() => {
        const handleHistoryChange = () => scrollTo();
        const removeHistoryListener = addHistoryListener(handleHistoryChange);

        return () => removeHistoryListener();
    }, [addHistoryListener]);

    if (renderError || configError || userDetailsError) {
        return (
            <HeadProvider>
                <Title>{`Home Page - ${STORE_NAME}`}</Title>
                <Main isMasked={true} />
                <Mask isActive={true} />
                <ToastContainer />
                <LoadingOverlay />
            </HeadProvider>
        );
    }

    if (!loaded) {
        return <LoadingOverlay />;
    }

    return (
        <HeadProvider>
            <Title>{`Home Page - ${STORE_NAME}`}</Title>
            <Switch>
                <Route path={rootCheckoutRoute} component={Checkout} />
                <Main isMasked={hasOverlay}>
                    <Routes />
                </Main>
            </Switch>
            <Mask isActive={hasOverlay} dismiss={handleCloseDrawer} />
            <CustomerModal
                handleClose={hideCustomerModal}
                modalType={customerModalType}
            />
            <MobileMenu />
            <ToastContainer />
            <LoadingOverlay />
            <CookieWarning />
        </HeadProvider>
    );
};

App.propTypes = {
    markErrorHandled: func,
    renderError: shape({
        stack: string
    }),
    unhandledErrors: array
};

export default App;
