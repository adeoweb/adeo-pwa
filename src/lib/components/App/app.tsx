import React, { useCallback, useEffect, Suspense } from 'react';
import AlertCircle from 'react-feather/dist/icons/alert-circle';
import CloudOff from 'react-feather/dist/icons/cloud-off';
import RefreshCcw from 'react-feather/dist/icons/refresh-ccw';
import Wifi from 'react-feather/dist/icons/wifi';

import { useToasts } from '@magento/peregrine';

import CustomerModal from 'src/lib/components/CustomerModal';
import LoadingOverlay from 'src/lib/components/LoadingOverlay';
import { Route, Switch, useHistory } from 'src/lib/drivers';
import SIGN_OUT_MUTATION from 'src/lib/queries/signOut.graphql';
import scrollTo from 'src/lib/util/scrollTo';
import { useApp } from 'src/peregrine/lib/talons/adeoweb/App/useApp';
import { useLoadConfig } from 'src/peregrine/lib/talons/adeoweb/App/useLoadConfig';
import { useLoadUser } from 'src/peregrine/lib/talons/adeoweb/User/useLoadUser';
import { useUserSession } from 'src/peregrine/lib/talons/adeoweb/User/useUserSession';

import { HTML_UPDATE_AVAILABLE } from '../../constants/swMessageTypes';
import GET_CUSTOMER_QUERY from '../../queries/getCustomer.graphql';
import GET_STORE_CONFIG_DATA from '../../queries/getStoreConfigData.graphql';
import { registerMessageHandler } from '../../util/swUtils';
import Checkout, { rootCheckoutRoute } from '../Checkout';
import CookieWarning from '../CookieWarning';
import { HeadProvider, Title } from '../Head';
import Icon from '../Icon';
import Main from '../Main';
import Mask from '../Mask';
import MobileMenu from '../MobileMenu';
import Routes from '../Routes';
import ToastContainer from '../ToastContainer';

require('../../styles/styles.scss');

const OnlineIcon = <Icon src={Wifi} attrs={{ width: 18 }} />;
const OfflineIcon = <Icon src={CloudOff} attrs={{ width: 18 }} />;
const ErrorIcon = <Icon src={AlertCircle} attrs={{ width: 18 }} />;
const UpdateIcon = <Icon src={RefreshCcw} attrs={{ width: 18 }} />;

const ERROR_MESSAGE = 'Sorry! An unexpected error occurred.';

interface IAppProps {
    markErrorHandled: () => void;
    renderError: {
        stack: string;
    };
    unhandledErrors: string[];
}

const App = (props: IAppProps): JSX.Element => {
    const { markErrorHandled, renderError, unhandledErrors } = props;
    const [, { addToast }] = useToasts();

    const handleIsOffline = useCallback(() => {
        addToast({
            type: 'error',
            icon: OfflineIcon,
            message: 'You are offline. Some features may be unavailable.',
            timeout: 3000
        });
    }, [addToast]);

    const handleIsOnline = useCallback(() => {
        addToast({
            type: 'info',
            icon: OnlineIcon,
            message: 'You are online.',
            timeout: 3000
        });
    }, [addToast]);

    const handleHTMLUpdate = useCallback(
        resetHTMLUpdateAvailableFlag => {
            addToast({
                type: 'warning',
                icon: UpdateIcon,
                message: 'Update available. Please refresh.',
                actionText: 'Refresh',
                timeout: 0,
                onAction: () => {
                    location.reload();
                },
                onDismiss: removeToast => {
                    resetHTMLUpdateAvailableFlag();
                    removeToast();
                }
            });
        },
        [addToast]
    );

    const handleError = useCallback(
        (_error, id, loc, handleDismissError) => {
            const errorToastProps = {
                icon: ErrorIcon,
                message: `${ERROR_MESSAGE}\nDebug: ${id} ${loc}`,
                onDismiss: remove => {
                    handleDismissError();
                    remove();
                },
                timeout: 15000,
                type: 'error'
            };

            addToast(errorToastProps);
        },
        [addToast]
    );

    const talonProps = useApp({
        handleError,
        handleIsOffline,
        handleIsOnline,
        handleHTMLUpdate,
        markErrorHandled,
        renderError,
        unhandledErrors
    });

    const {
        hasOverlay,
        handleCloseDrawer,
        setHTMLUpdateAvailable,
        hideCustomerModal,
        customerModalType
    } = talonProps;

    const { error: configError, loaded } = useLoadConfig({
        getConfigQuery: GET_STORE_CONFIG_DATA
    });

    const { error: userDetailsError } = useLoadUser({
        getUserDetailsQuery: GET_CUSTOMER_QUERY
    });

    useUserSession({ signOutMutation: SIGN_OUT_MUTATION });

    useEffect((): any => {
        const unregisterHandler = registerMessageHandler(
            HTML_UPDATE_AVAILABLE,
            () => {
                setHTMLUpdateAvailable(true);
            }
        );
        return unregisterHandler;
    }, [setHTMLUpdateAvailable]);

    useEffect(() => {
        document.body.classList.add('loaded');
    }, [loaded]);

    const { listen: addHistoryListener } = useHistory();

    useEffect(() => {
        const handleHistoryChange = () => scrollTo();
        const removeHistoryListener = addHistoryListener(handleHistoryChange);

        return () => removeHistoryListener();
    }, [addHistoryListener]);

    if (renderError || configError || userDetailsError) {
        return (
            <HeadProvider>
                <Suspense fallback={<LoadingOverlay />}>
                    <Title>{`Home Page - ${STORE_NAME}`}</Title>
                    <Main isMasked={true} />
                    <Mask isActive={true} />
                    <ToastContainer />
                    <LoadingOverlay />
                </Suspense>
            </HeadProvider>
        );
    }

    if (!loaded) {
        return <LoadingOverlay />;
    }

    return (
        <HeadProvider>
            <Suspense fallback={<LoadingOverlay />}>
                <Title>{`Home Page - ${STORE_NAME}`}</Title>
                <Switch>
                    <Route path={rootCheckoutRoute} component={Checkout} />
                    <Main isMasked={hasOverlay}>
                        <Routes />
                    </Main>
                </Switch>
                <Mask isActive={hasOverlay} dismiss={handleCloseDrawer} />
                <CustomerModal
                    handleClose={hideCustomerModal}
                    modalType={customerModalType}
                />
                <MobileMenu />
                <ToastContainer />
                <LoadingOverlay />
                <CookieWarning />
            </Suspense>
        </HeadProvider>
    );
};

export default App;
