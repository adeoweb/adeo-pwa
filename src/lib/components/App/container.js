import React from 'react';
import { IntlProvider } from 'react-intl';

import { useErrorContext } from 'src/peregrine/lib/context/adeoweb/unhandledErrors';
import { ApolloProvider as ApolloHooksProvider } from '@apollo/react-hooks';

import App from './app';
import { useErrorBoundary } from './useErrorBoundary';

const AppContainer = ({ apolloClient }) => {
    const ErrorBoundary = useErrorBoundary(App);
    const [unhandledErrors, errorApi] = useErrorContext();

    return (
        <ApolloHooksProvider client={apolloClient}>
            <IntlProvider locale="en">
                <ErrorBoundary
                    unhandledErrors={unhandledErrors}
                    {...errorApi}
                />
            </IntlProvider>
        </ApolloHooksProvider>
    );
};

export default AppContainer;
