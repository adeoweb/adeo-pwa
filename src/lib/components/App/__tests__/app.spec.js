import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';

import { createTestInstance } from '@magento/peregrine';

import { Switch } from 'src/lib/drivers';
import { useAppContext } from 'src/peregrine/lib/context/adeoweb/app';
import { useLoadConfig } from 'src/peregrine/lib/talons/adeoweb/App/useLoadConfig';

import Main from '../../Main';
import Mask from '../../Mask';
import MobileMenu from '../../MobileMenu';
import Routes from '../../Routes';

jest.mock('../../Head', () => ({
    HeadProvider: ({ children }) => <div>{children}</div>,
    Title: () => 'Title'
}));
jest.mock('../../Main', () => 'Main');
jest.mock('../../MobileMenu', () => 'MobileMenu');
jest.mock('../../Routes', () => 'Routes');
jest.mock('../../ToastContainer', () => 'ToastContainer');
jest.mock('owl.carousel', () => {});

const mockAddToast = jest.fn();
jest.mock('@magento/peregrine', () => {
    const useToasts = jest.fn(() => [
        { toasts: new Map() },
        { addToast: mockAddToast }
    ]);

    return {
        ...jest.requireActual('@magento/peregrine'),
        useToasts
    };
});

jest.mock('src/peregrine/lib/context/adeoweb/app', () => {
    const state = {
        hasBeenOffline: false,
        isOnline: true,
        overlay: false,
        drawer: null
    };
    const api = {
        closeDrawer: jest.fn()
    };
    const useAppContext = jest.fn(() => [state, api]);

    return { useAppContext };
});

jest.mock('src/peregrine/lib/context/adeoweb/checkout', () => {
    const state = {};
    const api = {
        actions: {
            reset: jest.fn()
        }
    };
    const useCheckoutContext = jest.fn(() => [state, api]);

    return { useCheckoutContext };
});

jest.mock('@magento/peregrine/lib/context/cart', () => {
    const state = {
        cartId: null
    };
    const api = {
        getCartDetails: jest.fn(),
        setCartId: id => {
            state.cartId = id;
        }
    };
    const useCartContext = jest.fn(() => [state, api]);

    return { useCartContext };
});

jest.mock('@magento/peregrine/lib/util/createErrorRecord', () => ({
    __esModule: true,
    default: jest.fn().mockReturnValue({
        error: { message: 'A render error', stack: 'errorStack' },
        id: '1',
        loc: '1'
    })
}));

jest.mock('@apollo/react-hooks', () => ({
    useMutation: jest.fn().mockImplementation(() => [
        jest.fn().mockImplementation(() => {
            return {
                data: {
                    createEmptyCart: 'cartIdFromGraphQL'
                }
            };
        })
    ]),
    useQuery: jest.fn(() => {
        return {
            data: {
                cmsBlocks: {
                    items: ['']
                }
            }
        };
    })
}));

jest.mock('src/peregrine/lib/talons/adeoweb/App/useLoadConfig', () => {
    const useLoadConfig = jest.fn(() => ({
        error: null,
        loaded: true
    }));

    return { useLoadConfig };
});

jest.mock('src/peregrine/lib/talons/adeoweb/User/useLoadUser', () => {
    const useLoadUser = jest.fn(() => ({
        error: null
    }));

    return { useLoadUser };
});

jest.mock('src/peregrine/lib/context/adeoweb/messageCard', () => {
    const useMessageCardContext = jest.fn(() => [
        {
            messageBlocks: {
                block: []
            }
        },
        { clearAllMessages: jest.fn(() => {}) }
    ]);

    return { useMessageCardContext };
});

jest.mock('src/peregrine/lib/talons/adeoweb/User/useUserSession', () => {
    return { useUserSession: jest.fn() };
});

jest.mock('src/lib/drivers', () => {
    const useHistory = jest.fn(() => ({
        listen: jest.fn(() => {})
    }));

    return {
        ...jest.requireActual('src/lib/drivers'),
        useHistory
    };
});

// require app after mock is complete
const App = require('../app').default;

const getAndConfirmProps = (parent, type, props = {}) => {
    const instance = parent.findByType(type);
    expect(instance.props).toMatchObject(props);
    return instance;
};

const { location } = window;

beforeAll(() => {
    global.STORE_NAME = 'Venia';
    delete window.location;
    window.location = { reload: jest.fn() };
});

afterAll(() => (window.location = location));

test('renders a full page with onlineIndicator and routes', () => {
    const [appState, appApi] = useAppContext();
    const mockedReturnValue = [
        {
            ...appState,
            drawer: '',
            overlay: false,
            hasBeenOffline: true,
            isOnline: false
        },
        appApi
    ];

    useAppContext.mockReturnValueOnce(mockedReturnValue);

    const appProps = {
        markErrorHandled: jest.fn(),
        unhandledErrors: []
    };
    const { root } = createTestInstance(
        <Router>
            <App {...appProps} />
        </Router>
    );

    getAndConfirmProps(root, MobileMenu);

    const switchComponent = getAndConfirmProps(root, Switch);

    expect(mockAddToast).toHaveBeenCalledWith({
        type: 'error',
        icon: expect.any(Object),
        message: 'You are offline. Some features may be unavailable.',
        timeout: 3000
    });
    // renderRoutes should just return a fake component here
    expect(switchComponent.findByType(Routes)).toBeTruthy();

    expect(switchComponent.findByType(Main)).toBeTruthy();

    const mask = getAndConfirmProps(root, Mask, {
        isActive: false,
        dismiss: expect.any(Function)
    });

    // appropriate positioning
    const {
        parent: { children: siblings }
    } = switchComponent;
    expect(siblings.indexOf(switchComponent)).toBeLessThan(
        siblings.indexOf(mask)
    );
});

test('checks if Main is in Switch', () => {});

test('displays onlineIndicator online if hasBeenOffline', () => {
    const [appState, appApi] = useAppContext();
    const mockedReturnValue = [
        {
            ...appState,
            drawer: '',
            overlay: false,
            hasBeenOffline: true,
            isOnline: true
        },
        appApi
    ];

    useAppContext.mockReturnValueOnce(mockedReturnValue);

    const appProps = {
        markErrorHandled: jest.fn(),
        unhandledErrors: []
    };

    createTestInstance(
        <Router>
            <App {...appProps} />
        </Router>
    );
    expect(mockAddToast).toHaveBeenCalledWith({
        type: 'info',
        icon: expect.any(Object),
        message: 'You are online.',
        timeout: 3000
    });
});

test('displays open nav', () => {
    const [appState, appApi] = useAppContext();
    useAppContext.mockReturnValueOnce([
        {
            ...appState,
            drawer: 'nav',
            overlay: false,
            hasBeenOffline: true,
            isOnline: true
        },
        appApi
    ]);
    const props = {
        markErrorHandled: jest.fn(),
        unhandledErrors: []
    };

    const { root: openNav } = createTestInstance(
        <Router>
            <App {...props} />
        </Router>
    );

    getAndConfirmProps(openNav, MobileMenu);
});

test('renders with renderErrors', () => {
    const appProps = {
        app: {
            drawer: '',
            overlay: false,
            hasBeenOffline: true,
            isOnline: false
        },
        closeDrawer: jest.fn(),
        markErrorHandled: jest.fn(),
        unhandledErrors: [],
        renderError: new Error('A render error!')
    };

    const root = createTestInstance(
        <Router>
            <App {...appProps} />
        </Router>
    );

    expect(root.toJSON()).toMatchSnapshot();
});

test('renders with unhandledErrors', () => {
    const appProps = {
        app: {
            drawer: '',
            overlay: false,
            hasBeenOffline: true,
            isOnline: false
        },
        closeDrawer: jest.fn(),
        markErrorHandled: jest.fn(),
        unhandledErrors: [{ error: new Error('A render error!') }],
        renderError: null
    };

    const root = createTestInstance(
        <Router>
            <App {...appProps} />
        </Router>
    );

    expect(root.toJSON()).toMatchSnapshot();
});

test('renders with configError', () => {
    const appProps = {
        app: {
            drawer: '',
            overlay: false,
            hasBeenOffline: true,
            isOnline: false
        },
        closeDrawer: jest.fn(),
        markErrorHandled: jest.fn(),
        unhandledErrors: [],
        renderError: null
    };
    useLoadConfig.mockReturnValueOnce({
        ...useLoadConfig(),
        error: 'config error'
    });

    const root = createTestInstance(
        <Router>
            <App {...appProps} />
        </Router>
    );

    expect(root.toJSON()).toMatchSnapshot();
});

test('adds no toasts when no errors are present', () => {
    const appProps = {
        app: {
            drawer: '',
            overlay: false,
            hasBeenOffline: false,
            isOnline: true
        },
        closeDrawer: jest.fn(),
        markErrorHandled: jest.fn(),
        unhandledErrors: [],
        renderError: null
    };

    createTestInstance(
        <Router>
            <App {...appProps} />
        </Router>
    );

    expect(mockAddToast).not.toHaveBeenCalled();
});

test('adds toasts for render errors', () => {
    const appProps = {
        app: {
            drawer: '',
            overlay: false,
            hasBeenOffline: true,
            isOnline: false
        },
        closeDrawer: jest.fn(),
        markErrorHandled: jest.fn(),
        unhandledErrors: [],
        renderError: new Error('A render error!')
    };

    createTestInstance(
        <Router>
            <App {...appProps} />
        </Router>
    );

    expect(mockAddToast).toHaveBeenCalledWith({
        icon: expect.any(Object),
        message: expect.any(String),
        onDismiss: expect.any(Function),
        timeout: expect.any(Number),
        type: 'error'
    });
});

test('adds toasts for unhandled errors', () => {
    const appProps = {
        app: {
            drawer: '',
            overlay: false,
            hasBeenOffline: true,
            isOnline: false
        },
        closeDrawer: jest.fn(),
        markErrorHandled: jest.fn(),
        unhandledErrors: [{ error: new Error('A render error!') }],
        renderError: null
    };

    createTestInstance(
        <Router>
            <App {...appProps} />
        </Router>
    );

    expect(mockAddToast).toHaveBeenCalledWith({
        icon: expect.any(Object),
        message: expect.any(String),
        onDismiss: expect.any(Function),
        timeout: expect.any(Number),
        type: 'error'
    });
});

test('displays update available message when a HTML_UPDATE_AVAILABLE message is received', () => {
    const appProps = {
        markErrorHandled: jest.fn(),
        unhandledErrors: []
    };

    createTestInstance(
        <Router>
            <App {...appProps} />
        </Router>
    );

    window.postMessage('HTML_UPDATE_AVAILABLE', '*');

    setTimeout(() => {
        expect(mockAddToast).toHaveBeenCalledWith({
            type: 'warning',
            icon: expect.any(Object),
            message: 'Update available. Please refresh.',
            actionText: 'Refresh',
            timeout: 0,
            onAction: expect.any(Function),
            onDismiss: expect.any(Function)
        });
    }, 1000);
});
