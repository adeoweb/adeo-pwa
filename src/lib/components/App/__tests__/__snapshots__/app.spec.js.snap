// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with configError 1`] = `
<div>
  Title
  <Main
    isMasked={true}
  />
  <button />
  <ToastContainer />
  <div
    className="loading-overlay"
  >
    <div
      className="bounce-loader"
    >
      <div
        className="bounce1"
      />
      <div
        className="bounce2"
      />
      <div
        className="bounce3"
      />
    </div>
  </div>
</div>
`;

exports[`renders with renderErrors 1`] = `
<div>
  Title
  <Main
    isMasked={true}
  />
  <button />
  <ToastContainer />
  <div
    className="loading-overlay"
  >
    <div
      className="bounce-loader"
    >
      <div
        className="bounce1"
      />
      <div
        className="bounce2"
      />
      <div
        className="bounce3"
      />
    </div>
  </div>
</div>
`;

exports[`renders with unhandledErrors 1`] = `
<div>
  Title
  <Main
    computedMatch={
      Object {
        "isExact": true,
        "params": Object {},
        "path": "/",
        "url": "/",
      }
    }
    isMasked={false}
    location={
      Object {
        "hash": "",
        "pathname": "/",
        "search": "",
      }
    }
  >
    <Routes />
  </Main>
  <button
    onClick={[Function]}
  />
  <MobileMenu />
  <ToastContainer />
  <div
    className="loading-overlay"
  >
    <div
      className="bounce-loader"
    >
      <div
        className="bounce1"
      />
      <div
        className="bounce2"
      />
      <div
        className="bounce3"
      />
    </div>
  </div>
  <div
    className="warningContainer"
  >
    <div
      className="warningWrapper container"
    >
      <div
        className="textWrapper"
      >
        <div>
          <div />
        </div>
      </div>
      <button
        className="button btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Accept
      </button>
    </div>
  </div>
</div>
`;
